### ISSUE TITLE
Bug with xmlserializer when deployed to IIS

### ISSUE THREAD

**Issue by kostasxagoraris** [dotnet/runtime#44995]
I have a project that was created originally with blazor web assembly 3.2. 
It used some more .net standard 2.0 projects to serialize a few classes from and to xml (also under .net standard 2.0).
I recently upgraded the blazor web app to .net 5.0 . (the projects under .net standard 2.0 remained as they were)
While everything on my development machines run smoothly when i deploy to iis I cant instantiate the xml serializer

 [NotSupportedException] : "Parent does not have a default constructor. The default constructor must be explicitly defined."  

The stack trace
at System.Reflection.Emit.TypeBuilder.DefineDefaultConstructor(MethodAttributes attributes)
blazor.webassembly.js:1    at System.Xml.Serialization.XmlSerializationILGen.GenerateBaseSerializer(String baseSerializer, String readerClass, String writerClass, CodeIdentifiers classes)
blazor.webassembly.js:1    at System.Xml.Serialization.TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace)
blazor.webassembly.js:1    at System.Xml.Serialization.TempAssembly..ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location)
blazor.webassembly.js:1    at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace, String location)
blazor.webassembly.js:1    at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace)
blazor.webassembly.js:1    at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
blazor.webassembly.js:1    at System.Xml.Serialization.XmlSerializer..ctor(Type type)
blazor.webassembly.js:1    at PanasoftUtilities.XMLStatic.Create_ClassObject_From_XMLDeserialization[AADEBooksParam](String xml, String& Errormessage, Type[] ExtraTypes)

Source: System.Private.CoreLib

TargetSite: DefineDefaultConstructor

Of course there is no lack of default constructor anywhere. And anyway it works fine when running the site on visual studio and it worked fine under blazor 3.2 .net core 3.1.

I tried publishing the site using different combinations of x64 x86 and target runtime.  
I ve being struggling with the issue for 2 days. I don't know what else to do.

Do you need me to upload the published artifacts so that you have a look?

Thanks for your advice.

**Comment by javiercn** [comment]
This seems to be an issue with the XML serializer running on WASM, the user disabled trimming and the problem still persisted

**Comment by kostasxagoraris** [comment]
I also tried to copy the files from the build directory to a seperate web site on the same iis that i use for my development and i had the same exact issue.

**Comment by kostasxagoraris** [comment]
Yet another similar issue
Trying to transform xml with xslt
Here is the working code .net code 3.1

 protected async override Task OnParametersSetAsync()
    {
        using (StringReader text = new StringReader(TraceLog.Trace))
        {


            XPathDocument myXPathDoc = new XPathDocument(text);
            XslCompiledTransform myXslTrans = new XslCompiledTransform();
            string xmlStr = await Http.GetTraceXSLT(Navigation.ToAbsoluteUri("XSLTTraceToHTML.xslt").AbsoluteUri);
            using (StringReader str = new StringReader(xmlStr))
            using (var reader = System.Xml.XmlReader.Create(str))
            {
                myXslTrans.Load(reader);
                StringWriter myWriter = new StringWriter();
                myXslTrans.Transform(myXPathDoc, null, myWriter);
                TraceHtml = myWriter.GetStringBuilder().ToString();

            }
        }

    }

And here is the error with .Net 5.0


(unknown) Unhandled Exception:
window.console.error
(unknown) System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, con
window.console.error
(unknown)    at System.Reflection.Emit.CustomAttributeBuilder.Initialize(ConstructorInfo con, Object[] constructorArgs, PropertyInfo[] namedProperties, Object[] propertyValues, FieldInfo[] namedFields, Object[] fieldValues)
window.console.error
(unknown)    at System.Reflection.Emit.CustomAttributeBuilder..ctor(ConstructorInfo con, Object[] constructorArgs)
window.console.error
(unknown)    at System.Xml.Xsl.IlGen.XmlILModule.CreateLREModule()
window.console.error
(unknown)    at System.Xml.Xsl.IlGen.XmlILModule..cctor()
window.console.error
(unknown) crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: TypeInitialization_Type, System.Xml.Xsl.IlGen.XmlILModule
System.TypeInitializationException: TypeInitialization_Type, System.Xml.Xsl.IlGen.XmlILModule
 ---> System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, con
   at System.Reflection.Emit.CustomAttributeBuilder.Initialize(ConstructorInfo con, Object[] constructorArgs, PropertyInfo[] namedProperties, Object[] propertyValues, FieldInfo[] namedFields, Object[] fieldValues)
   at System.Reflection.Emit.CustomAttributeBuilder..ctor(ConstructorInfo con, Object[] constructorArgs)
   at System.Xml.Xsl.IlGen.XmlILModule.CreateLREModule()
   at System.Xml.Xsl.IlGen.XmlILModule..cctor()
   Exception_EndOfInnerExceptionStack
   at System.Xml.Xsl.XmlILGenerator.Generate(QilExpression query, TypeBuilder typeBldr)
   at System.Xml.Xsl.XslCompiledTransform.CompileQilToMsil(XsltSettings settings)
   at System.Xml.Xsl.XslCompiledTransform.LoadInternal(Object stylesheet, XsltSettings settings, XmlResolver stylesheetResolver)
   at System.Xml.Xsl.XslCompiledTransform.Load(XmlReader stylesheet)
   at Panasoft.Web.MyData.Client.Pages.Monitor.TraceView.OnParametersSetAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)

**Comment by danmosemsft** [comment]
@eerhardt I wonder whether it would be useful to create a meta label to add to any trim safety issues.

**Comment by eerhardt** [comment]
> I wonder whether it would be useful to create a meta label to add to any trim safety issues.

I've been using the `linkable-framework` label.

The original issue is a duplicate of #41389 and will be resolved with #44772 once it gets approved.

The [2nd issue](https://github.com/dotnet/runtime/issues/44995#issuecomment-732819444) is a new one, it looks like the following code needs to be annotated correctly for the ILLinker to preserve the constructors correctly:

https://github.com/dotnet/runtime/blob/aebd598664e7d931eee19eb1779c0ace5d1c3600/src/libraries/System.Private.Xml/src/System/Xml/Xsl/IlGen/GenerateHelper.cs#L135-L159

Side-note: The ILLink warnings caught this as well:

https://github.com/dotnet/runtime/blob/aebd598664e7d931eee19eb1779c0ace5d1c3600/src/libraries/System.Private.Xml/src/ILLink/ILLink.Suppressions.xml#L130-L141

**Comment by eerhardt** [comment]
@kostasxagoraris - to try to get you unblocked on `net5.0`, can you try the following?

In your Blazor Client.csproj, add the following lines:

```xml
  <ItemGroup>
    <TrimmerRootAssembly Include="System.Private.Xml" />
    <TrimmerRootDescriptor Include="LinkerWorkaround.xml" />
  </ItemGroup>
```

And then create a file named `LinkerWorkaround.xml` in your Blazor Client.csproj folder with the following contents:

```xml
<?xml version="1.0" encoding="utf-8" ?>
<linker>
  <assembly fullname="System.Private.CoreLib">
    <type fullname="System.Security.SecurityTransparentAttribute" />
  </assembly>
</linker>
```

I tried this myself, and it worked for me. Let me know if it works for you. We will get this fixed in net6.0.

**Comment by kostasxagoraris** [comment]
@eerhardt Thank you very much That solved all issues. Now it works smoothly!!
